files/grid-studies-2-2.py

```diff
@@ -7,48 +7,59 @@ class GridStudies(monome.GridApp):
     def __init__(self):
         super().__init__()
 
+    width = 0
+    height = 0
+    step = [[0 for col in range(16)] for row in range(16)]
+    play_position = -1
+
+    # when grid is plugged in via USB:
     def on_grid_ready(self):
-        self.step = [[0 for col in range(self.grid.width)] for row in range(6)]
-        self.play_position = 0
+        self.width = self.grid.width
+        self.height = self.grid.height
+        self.sequencer_rows = self.height-2
+        self.connected = True
 
-        asyncio.async(self.play())
+    # when grid is physically disconnected:
+    def on_grid_disconnect(self,*args):
+        self.connected = False
 
     async def play(self):
         while True:
-            if self.play_position == self.grid.width - 1:
+            await asyncio.sleep(0.1)
+            
+            if self.play_position == self.width - 1:
                 self.play_position = 0
             else:
                 self.play_position += 1
-
-            self.draw()
-
-            await  asyncio.sleep(0.1)
+ 
+            if self.connected:
+                self.draw()
```

Comment:
in this loop i'd recommend drawing before awaiting for `sleep` so the grid gets rendered immediately on application start.

Reply:
oh! i swapped the order of these and the sequencer doesn't draw or progress at all on application start. i guess i was also relying on norns clock conventions, where we typically sync before progressing variables + redrawing. lmk if i'm totally missing something tho!

Reply:
ah, sorry. i forgot that `self.draw()` is also called in `on_grid_ready`, so this shouldn't be a problem! looks good to go!

Reply:
for context: basically my concern was that the pause in the beginning of `play()` will cause a delay when rendering the grid - i usually do a grid refresh, followed by "sync" call before the loop in sequencing coroutines as follows (in a kind of pseudocode):

```
async def play():
    self.play_position = -1 # will become 0 soon anyway
    self.draw()
    await clock.sync(1/4)

    while True:
        self.play_position += 1

        # triggers, etc. go here
 
        self.draw()
        await clock.sync(1/4)
```

so the code is logically divided into 1) updating position 2) sending out triggers 3) draw+sync parts.

but i think it's fine either way for the purpose of this tutorial - the very first grid rendering can happen in `on_grid_ready` too.

Reply:
ah gotcha, that's super clear.
i really appreciate your time + insights -- thank you for the fantastic python lessons via these comments!!